File: 15-3sum/3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> ans;
        unordered_map<int, int> mp;
        int n = nums.size();
        for(int i = 0; i < nums.size(); i++)
            mp[nums[i]] = i;
        sort(nums.begin(), nums.end());
        for(int i=0; i<n; i++){
            // while(i < n-2 && nums[i] == nums[i+1]) i++;
            if(i > 0 && nums[i] == nums[i-1]) continue;
            int j = i + 1, k = nums.size() - 1;
            while(j < k){
                int sum = nums[i] + nums[j] + nums[k];
                
                
                if(sum == 0){
                    ans.push_back({nums[i],nums[j],nums[k]});
                    while(j < k && nums[j] == nums[j+1]) j++;
                    j++;
                    while(k > j && nums[k] == nums[k-1]) k--;
                    k--;
                }else if(sum < 0){
                    j++;
                }else{
                    k--;
                }
                
            }
        }
        return ans;
    }
};

File: 15-3sum/3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        int n = nums.length;
        Arrays.sort(nums);
        Set<List<Integer>> ans = new HashSet<>();
        for(int i = 0; i < n - 2; i++){
            int j = i + 1;
            int k = n - 1;
            while(j < k){
                int sum = nums[i] + nums[j] + nums[k];
                if(sum == 0){
                    ans.add(Arrays.asList(nums[i], nums[j], nums[k]));
                    while(j < k && nums[j] == nums[j+1]){
                        j++;
                    }
                    while(k > j && nums[k] == nums[k-1]){
                        k--;
                    }
                    j++;
                    k--;
                }else if(sum < 0){
                    j++;
                }else{
                    k--;
                }
                
            }
        }
        return new ArrayList<>(ans);
    }
}